phoenix-core/src/main/java/org/apache/phoenix/expression/LiteralExpression.java

```diff
@@ -338,6 +338,17 @@ public PDataType getDataType() {
 
     @Override
     public Integer getMaxLength() {
+        // For literals representing arrays of CHAR or BINARY, the byte size is null and the max
```

Comment:
Isn't the logic from 344 to 350 already covered by public static LiteralExpression newConstant(...)? It seems that the value/maxLength are assigned by newConstant static function  in which a LiteralExpression object is created.

Reply:
I ran my debugger through this and it looks like [that code](https://github.com/apache/phoenix/blob/master/phoenix-core/src/main/java/org/apache/phoenix/expression/LiteralExpression.java#L206) is never hit. Overall, this class needs to be refactored to follow the builder pattern. There are 5 constructors and 10 overloaded methods for `newConstant` which basically instantiates a new `LiteralExpression`. I am not aware of all of the code paths involved here and whether the `newConstant` method where this logic already exists is getting called in all those cases. As I mentioned, this code is not getting hit for the test cases I've added in this PR, so I don't think it is safe to rely on changing `newConstant`.

Reply:
The interface of LiteralExpression is a mess. While it provides newConstant() APIs to build the object, it also provide two public constructors.

Overall, it seems to me that all production and test code paths should call newConstant() APIs to build the LiteralExpression object. Once value/type/byteValue/maxLength are passed through the following private LiteralExpression() constructor, they should never be change and the get accessor should return exactly what is. Having this ad-hoc change in getMaxLength() will make code quality to be worse and hurt the performance of this getter accessor. If this introduce big refactor which is out of scope of this JIRA PHOENIX-5416. Shall we leave getMaxLength() as it was and open a new JIRA to refactor LiteralExpression class and its callers? 

private LiteralExpression(Object value, PDataType type, byte[] byteValue,
            Integer maxLength, Integer scale, SortOrder sortOrder, Determinism deterministic) {